### PR TITLE
fix(*): localhost exposed application shouldn't be reachable

### LINKED ISSUE #4630: Applications bound to localhost behind the Service are reachable when Kuma deployed
### What happened?

When the application running within a `Pod` is bind to the localhost, has a Service that exposes it and we deploy `Kuma`, you can reach the application. But the request to the application bind to podIP is failing. We should be more consistent with migration for users and increase security by not exposing localhost bound applications.

Step to reproduce:
1. Deploy Kuma
2. Deploy demo application bound to localhost and expose it by the Service
3. Install another pod with curl/wget
4. Request `CLUSTER_IP:PORT` from the client pod
5. request will pass

Step to reproduce case with pod IP:
1. Deploy Kuma
2. Deploy demo application bound to podIP and expose it by the Service
3. Install another pod with curl/wget
4. Request `CLUSTER_IP:PORT` from the client pod
5. Request will fail

Deployment without the mesh has different behavior and request to application bound to localhost fail while to podIP its success.

### PR DESCRIPTION
**Pull request by lukidzi**
### Summary
When a user deploys Kuma the behavior of communication with services changes. 
When there is no Kuma and the application running within the pod is binding to PodIP or Wildcard(`0.0.0.0`) other pods are able to reach the application. After Kuma is deployed it's not possible anymore. Instead, applications that bind to `localhost` or wildcard are exposed outside. That's a big security threat.  In this PR we are introducing a different way how the inbound cluster is configured by default.

Previously we had an inbound listener that had a static cluster that pointed to the address `localhost:PORT`. Now it's going to point to `DataplaneIP`
1. Application that is binding to `localhost` won't be accessible anymore. 
2. User can expose application binding to `localhost` by `dataplane.networking.inbound[].serviceAddress`
 
#### What can I do to make the smooth upgrade?
- check if your applications are listening on `localhost` and should be exposed outside - if no change to `0.0.0.0`
- if you want to expose application binding to `localhost` set `dataplane.networking.inbound[].serviceAddress` to `127.0.0.1`
- you can disable this behavior with `kuma-cp` configuration or env to `kuma-cp` `KUMA_DEFAULTS_ENABLE_LOCALHOST_INBOUND_CLUSTERS` - not recommended, we are going to remove this flag in the future

### Full changelog

* [Changed the default inbound cluster IP from `localhost` to `DataplaneIP`]
* [Setting `upstream_bind_address` for inbound clusters that don't set have localhost address]
* [Added e2e tests that check behavior with different application bindings]
* [Changed inbound clusters naming: `localhost` -> `inbound`]
* [Made changes to applications metrics scraper to use `DataplaneIP`]

- [X] Link to docs PR or issue -- none
- [X] Link to UI issue or PR -- none
- [X] Is the [issue worked on linked][1]? -- Fix #4630 
- [X] Unit Tests --
- [X] E2E Tests --
- [X] Manual Universal Tests --
- [X] Manual Kubernetes Tests --
- [X] Do you need to update [`UPGRADE.md`](/UPGRADE.md)? --
- [X] Does it need to be backported according to the [backporting policy](/CONTRIBUTING.md#backporting)? -- none

### CODE REVIEW

File: pkg/core/bootstrap/bootstrap.go [hunk 1]
```diff
@@ -125,6 +126,9 @@ func buildRuntime(appCtx context.Context, cfg kuma_cp.Config) (core_runtime.Runt
 		))
 	}
 
+	// The setting should be removed, and there is no easy way to set it without breaking most of the code
+	mesh_proto.EnableLocalhostInboundClusters = builder.Config().Defaults.EnableLocalhostInboundClusters
```
Comment (michaelbeaumont) [inline pkg/core/bootstrap/bootstrap.go:130]: Would it be possible to add this setting to the `Dataplane` proto and then, based on the CP setting, setting the field when a `Dataplane` is created?

File: pkg/xds/generator/inbound_proxy_generator.go [hunk 2]
```diff
@@ -36,10 +36,15 @@ func (g InboundProxyGenerator) Generate(ctx xds_context.Context, proxy *core_xds
 		protocol := core_mesh.ParseProtocol(iface.GetProtocol())
 
 		// generate CDS resource
-		localClusterName := envoy_names.GetLocalClusterName(endpoint.WorkloadPort)
+		localClusterName := envoy_names.GetInboundClusterName(endpoint.WorkloadPort)
 		clusterBuilder := envoy_clusters.NewClusterBuilder(proxy.APIVersion).
 			Configure(envoy_clusters.ProvidedEndpointCluster(localClusterName, false, core_xds.Endpoint{Target: endpoint.WorkloadIP, Port: endpoint.WorkloadPort})).
 			Configure(envoy_clusters.Timeout(defaults_mesh.DefaultInboundTimeout(), protocol))
+		// localhost traffic is routed dirrectly to the application, in case of other interface we are going to set source address to
+		// 127.0.0.6 to avoid redirections and thanks to first iptables rule just return fast
```
Comment (michaelbeaumont) [inline pkg/xds/generator/inbound_proxy_generator.go:44]: No idea if relevant at this point, more from curiosity but does this break applications that listen on `127.0.0.6`? It's a loopback address just like `127.0.0.1`.

File: test/e2e_env/kubernetes/gateway/cross-mesh.go [hunk 3]
```diff
@@ -134,6 +134,7 @@ func CrossMeshGatewayOnKubernetes() {
 	E2EAfterAll(func() {
 		Expect(env.Cluster.TriggerDeleteNamespace(gatewayClientNamespaceOtherMesh)).To(Succeed())
 		Expect(env.Cluster.TriggerDeleteNamespace(gatewayClientNamespaceSameMesh)).To(Succeed())
+		Expect(env.Cluster.TriggerDeleteNamespace(gatewayClientOutsideMesh)).To(Succeed())
```
Comment (michaelbeaumont) [inline test/e2e_env/kubernetes/gateway/cross-mesh.go:137]: Nice catch.

File: test/server/cmd/echo.go [hunk 4]
```diff
@@ -78,12 +79,13 @@ func newEchoHTTPCmd() *cobra.Command {
 				})
 			}
 			if args.tls {
-				return http.ListenAndServeTLS(fmt.Sprintf(":%d", args.port), args.crtFile, args.keyFile, nil)
+				return http.ListenAndServeTLS(fmt.Sprintf("%s:%d", args.ip, args.port), args.crtFile, args.keyFile, nil)
 			}
-			return http.ListenAndServe(fmt.Sprintf(":%d", args.port), nil)
+			return http.ListenAndServe(fmt.Sprintf("%s:%d", args.ip, args.port), nil)
```
Comment (michaelbeaumont) [inline test/server/cmd/echo.go:84]: Better `net.JoinHostPort`?

**Review by michaelbeaumont — COMMENTED** [review]
Nice, thanks for the commits, much easier to review IMO

**Review by lobkovilya — APPROVED** [review]

File: pkg/xds/generator/inbound_proxy_generator.go (hunk 2 above)
Comment (lukidzi) [reply pkg/xds/generator/inbound_proxy_generator.go:44]: I don't think it should break. Envoy should override `sourceIp` and not a destination. Also, we do not bind to this address.

File: pkg/core/bootstrap/bootstrap.go (hunk 1 above)
Comment (lukidzi) [reply pkg/core/bootstrap/bootstrap.go:130]: It's possible but later might be problematic. This field should be removed in the future if we introduce a new field in the meanwhile why might need to keep the order and the field.

File: test/server/cmd/echo.go [hunk 5]
```diff
@@ -79,9 +81,9 @@ func newEchoHTTPCmd() *cobra.Command {
 				})
 			}
 			if args.tls {
-				return http.ListenAndServeTLS(fmt.Sprintf("%s:%d", args.ip, args.port), args.crtFile, args.keyFile, nil)
+				return http.ListenAndServeTLS(net.JoinHostPort(args.ip, strconv.FormatInt(int64(args.port), 10)), args.crtFile, args.keyFile, nil)
```
Comment (michaelbeaumont) [inline test/server/cmd/echo.go:84]: fyi there's `strconv.Itoa`.

File: api/mesh/v1alpha1/dataplane_helpers.go [hunk 6]
```diff
@@ -12,6 +12,8 @@ import (
 	"github.com/pkg/errors"
 )
 
+var EnableLocalhostInboundClusters bool
```
[the remark this answers is not in the document]
Comment (michaelbeaumont) [reply api/mesh/v1alpha1/dataplane_helpers.go:17]: https://github.com/kumahq/kuma/issues/4772

**Review by michaelbeaumont — APPROVED** [review]

**Comment by lobkovilya** [comment]
@lukidzi what happens if application binds to `127.0.0.6` when acts as a client? Does it bypass the envoy in that case?

**Comment by slonka** [comment]
@lukidzi - you need to pull master for the `ci/run-full-matrix` label to work

**Comment by lukidzi** [comment]
> @lukidzi what happens if application binds to `127.0.0.6` when acts as a client? Does it bypass the envoy in that case?

It seems so, but it looks like it's the same behavior now. We've already had that rule so from iptables there is no change.

File: pkg/xds/generator/utils.go [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package generator
+
+import "net"
+
+func IsAddressIPv6(address string) bool {
```
Comment (jakubdyszkiewicz) [inline pkg/xds/generator/utils.go:5]: could we move it to pkg/utils/net

**Pull request merged**